internal/aws/containerinsight/httpclient/httpclient.go

```diff
@@ -0,0 +1,131 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpClient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"math"
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	defaultMaxRetries              = 3
+	defaultTimeout                 = 1 * time.Second
+	defaultBackoffRetryBaseInMills = 200
+	maxHttpResponseLength          = 5 * 1024 * 1024 // 5MB
+)
+
+type HttpClient struct {
+	maxRetries              int
+	backoffRetryBaseInMills int
+	client                  clientProvider
+}
+
+type HttpClientProvider interface {
+	Request(path string, ctx context.Context, logger *zap.Logger) ([]byte, error)
+}
+
+type clientProvider interface {
+	Do(request *http.Request) (*http.Response, error)
+}
+
+func New() HttpClientProvider {
+
+	client := &http.Client{Timeout: defaultTimeout}
+
+	return NewHttp(defaultMaxRetries, defaultBackoffRetryBaseInMills, client)
+
+}
+
+func NewHttp(maxRetries int, backoffRetryBaseInMills int, provider clientProvider) HttpClientProvider {
+	httpClient := &HttpClient{
+		maxRetries:              maxRetries,
+		backoffRetryBaseInMills: backoffRetryBaseInMills,
+		client:                  provider,
+	}
+
+	return httpClient
+}
+
+func (h *HttpClient) backoffSleep(currentRetryCount int) {
+	backoffInMillis := int64(float64(h.backoffRetryBaseInMills) * math.Pow(2, float64(currentRetryCount)))
+	sleepDuration := time.Millisecond * time.Duration(backoffInMillis)
+	if sleepDuration > 60*1000 {
+		sleepDuration = 60 * 1000
+	}
+	time.Sleep(sleepDuration)
+}
+
+func (h *HttpClient) Request(endpoint string, ctx context.Context, logger *zap.Logger) (body []byte, err error) {
+	for i := 0; i < h.maxRetries; i++ {
+		body, err = h.request(endpoint, ctx)
+		if err != nil {
```

Comment:
Not all errors should be retried.  This needs to have some logic for determining which errors are safe to retry and which will never succeed and should result in immediate termination.

Reply:
Changed it using new retryablehttp package.